Add DTO projection support via projectAs() query method

dereuromark opened the pull request:
## Summary
- Add `projectAs()` method on SelectQuery to project results directly into readonly DTOs instead of Entity objects
- Support two DTO creation styles: DtoMapper (reflection-based) and `createFromArray()` factory method
- Add `#[CollectionOf]` attribute for specifying collection element types
- Handle nested association hydration into DTOs

## Usage

```php
// Simple DTO projection
$articles = $articlesTable->find()
    ->projectAs(ArticleDto::class)
    ->toArray();

// With associations
$articles = $articlesTable->find()
    ->contain(['Authors', 'Comments'])
    ->projectAs(ArticleDto::class)
    ->toArray();
```

DTO classes can use either constructor parameters (mapped via reflection):
```php
readonly class ArticleDto {
    public function __construct(
        public int $id,
        public string $title,
        public ?AuthorDto $author = null,
        #[CollectionOf(CommentDto::class)]
        public array $comments = [],
    ) {}
}
```

Or a `createFromArray()` factory method for full control over mapping:
```php
readonly class ArticleDto {
    public function __construct(
        public int $id,
        public string $title,
        public ?AuthorDto $author = null,
    ) {}
    
    public static function createFromArray(array $data, bool $ignoreMissing = false): self {
        return new self(
            id: $data['id'],
            title: $data['title'],
            author: isset($data['author']) 
                ? AuthorDto::createFromArray($data['author'], $ignoreMissing)
                : null,
        );
    }
}
```

## Performance
DTOs use approximately 3x less memory than Entity objects while providing similar hydration speed. The `createFromArray()` method is ~2.5x faster than Entity hydration.

Refs https://github.com/cakephp/cakephp/issues/19131

dereuromark committed:
Add DTO projection support via projectAs() query method

Adds ability to project ORM query results directly into readonly DTOs
instead of Entity objects. This provides a more memory-efficient and
immutable way to fetch read-only data.

New features:
- projectAs() method on SelectQuery to specify a DTO class
- DtoMapper using PHP 8 reflection with named constructor arguments
- CollectionOf attribute for specifying array collection element types
- Support for createFromArray() factory method (cakephp-dto style)
- Nested association hydration into DTOs

dereuromark committed:
Fix phpstan error with class-string type assertions

markstory commented on src/ORM/Attribute/CollectionOf.php:
```diff
@@ -0,0 +1,50 @@
+ *     public function __construct(
+ *         public int $id,
+ *         public string $title,
+ *         #[CollectionOf(CommentDto::class)]
```
This will be a good test to find out how the community likes APIs using attributes.

markstory commented on src/ORM/ResultSetFactory.php:
```diff
@@ -242,6 +249,45 @@ protected function groupResult(array $row, array $data): EntityInterface|array
+     */
+    protected function hydrateDto(array $row, string $dtoClass): object
+    {
+        // Check for array style static factory method
+        if (method_exists($dtoClass, 'createFromArray')) {
+            return $dtoClass::createFromArray($row, true);
+        }
+
+        // Use DtoMapper for plain readonly DTOs with named constructor params
+        return $this->getDtoMapper()->map($row, $dtoClass);
```
This is a nice way to give developers an extension point to optimize things if they need it. Should there be an interface for `createFromArray` or is a method check good enough?

dereuromark replied:
I thought about it, but didnt find a way to make this possible on existing libraries out there, they might not have it (e.g. mine) and requiring it could be cumbersome to match up. The method exists check could be sufficient here.

dereuromark replied:
Do we want an interface here for cleaner code? Cake plugins like CakeDto can be adjusted here, for others we would need the custom factory pattern described below though, or the classic way of using formatResults().

markstory replied:
I think a `method_exists` call is probably good enough to get started. I don't think we would have much fun with an interface as the return type is going to be `mixed` or a pain to deal with.

rochamarcelo commented on src/ORM/DtoMapper.php:
```diff
@@ -0,0 +1,192 @@
+        $type = $param->getType();
+
+        $info = [
+            'name' => $param->getName(),
```
Could we have  a new  object (ParamInfo) here instead of array

dereuromark replied:
I think the idea was to avoid the object overhead for pure internal handling. 
If there is a chance this goes also outside maybe we would need it thought.

rochamarcelo commented on src/ORM/Attribute/CollectionOf.php:
````diff
@@ -0,0 +1,50 @@
+ * ```
+ */
+#[Attribute(Attribute::TARGET_PARAMETER | Attribute::TARGET_PROPERTY)]
+class CollectionOf
````
I think it is good to place attributes inside Attribute namespace, Cake\ORM\Attribute instead of  Cake\ORM

ADmad commented:
How will fields like `_joinData` which are auto added to entities for BTM association be handled when using a DTO? Will they be dropped/ignored?

dereuromark commented:
They will be mapped if present in the dto specs as property. I will check if there is a test.

ADmad commented:
```php
readonly class ArticleDto {
    public function __construct(
        public int $id,
        public string $title,
        public ?AuthorDto $author = null,
        #[CollectionOf(CommentDto::class)]
        public array $comments = [],
    ) {}
}
```

This looks similar to how an entity would look for the entity implementation I have made https://github.com/ADmad/cakephp-entity/blob/master/tests/test_app/TestApp/Model/Entity/Article.php

```php
<?php
declare(strict_types=1);

namespace TestApp\Model\Entity;

use ADmad\Entity\Datasource\Entity;
use BackedEnum;
use DateTimeInterface;

class Article extends Entity
{
    protected ?int $id;
    protected ?int $author_id;
    protected $user_id;
    protected string $title;
    protected ?string $body;
    protected string|BackedEnum $published;
    protected ?DateTimeInterface $created;
    protected ?DateTimeInterface $modified;
    protected ?DateTimeInterface $updated;
    protected ?DateTimeInterface $date_specialed;
    protected array $tags;
    protected ?Author $author;
    protected array $comments;
    protected array $unaproved_comments;
    protected $article;
    protected $not_in_schema;

    protected $comments_notype;
    protected $user;

    public function isRequired(): bool
    {
        return true;
    }
}
```

rochamarcelo commented on src/ORM/Attribute/CollectionOf.php:
```diff
@@ -0,0 +1,50 @@
+        public string $class,
+    ) {
+    }
+}
```
Could we have a simple attribute with all possible configs? Here is an example with name and collectionOf, both are optional

```
 ```php
  readonly class ArticleDto {
      public function __construct(
          public int $id,
          #[FieldDto(name: "main_title")] public string $title,
          #[FieldDto(collectionOf: CommentDto::class)]
          public array $comments = [],
      ) {}
  } ```

dereuromark replied:
Isnt that bit of scope creep? Renaming the fields is a bit of different concern
Usually thats handled in the DTOs using from/to kind of mapping configs.

markstory replied:
Doing attribute aliasing/inflection is a good scenario for `createFromArray()` as well.

dereuromark replied:
What do you propose here @markstory ?
Maybe you can clarify

ADmad replied:
> Renaming the fields is a bit of different concern

Being able to rename fields would be quite handy for tasks like report generation. For e.g. being able to represent `author.name` as `author_name` in xls/csv rendering.

dereuromark replied:
OK, what would be the naming/signature be then for this attribute?
If we brainstorm on full specs now first, I can finish this up in one batch.

And:
 Should I implement this as a replacement for #[CollectionOf], or keep both (with #[CollectionOf] as shorthand)?

markstory replied:
> Maybe you can clarify

My thought here is that we build provide simple constructor support, and if userland code need attribute renames, inflections or anything along those lines they can implement `createFromArray()`. We can add more attributes and more 'configuration' to DTO mapping later when we better understand the gaps/limitations in our simple solution.

dereuromark replied:
OK, so we address this later if needed.

dereuromark committed:
Move CollectionOf attribute to Cake\ORM\Attribute namespace

dereuromark committed:
Fix DtoMapper to pass through objects instead of mapping them

DateTime and other objects type-hinted on constructor parameters were
incorrectly being passed to map() as if they were arrays for nested DTO
mapping. Added is_array() check to only map actual arrays.

Added test cases for DateTime object handling.

dereuromark committed:
Move DTO projection to final formatter phase

Previously, DTO mapping happened in ResultSetFactory::groupResult()
during hydration, BEFORE formatters ran. This caused issues with
behaviors like TagBehavior that add formatters expecting arrays/entities.

Now DTO projection:
1. Returns arrays from groupResult() when projectAs() is used
2. Applies DTO mapping as a FINAL formatter after all others
3. Allows behavior formatters to work with arrays/entities first

This aligns with the user-space approach suggested in PR review:
disableHydration()->formatResults() but built-in for convenience.

dereuromark committed:
Restore createFromArray support in DTO projection

Use hydrateDto() (now public) instead of getDtoMapper()->map()
directly so that DTOs with static createFromArray() factory methods
(cakephp-dto style) continue to work.

dereuromark commented:
@ADmad 
The sandbox examples show it live if run locally:
https://github.com/dereuromark/cakephp-sandbox/pull/102/changes/b5ece8526cfd1aa506b0501f4f134fae514c8232
ignored unless you set a field for it, then it will be included as array/DTO:

Option 1: ?array (simple)
```php
  readonly class UserWithMatchingDto {
      public function __construct(
          public int $id,
          public string $username,
          public ?array $_matchingData = null,
      ) {}
  }
```
Option 2: ?MatchingDataDto (full type safety)
```php
  readonly class MatchingDataDto {
      public function __construct(
          public ?SimpleRoleDto $Roles = null,  // Property name = association alias
      ) {}
  }

  readonly class UserWithMatchingDtoTyped {
      public function __construct(
          public int $id,
          public string $username,
          public ?MatchingDataDto $_matchingData = null,
      ) {}
  }
```

dereuromark commented:
To support a wider range of DTO libs, we might want to create a DtoRegistry or some Configure approach on how to map them. 
Or simply allow a factory to be passed:
```php
  // For Valinor (external mapper)
  ->projectAs(UserDto::class, fn($data) => $valinorMapper->map(UserDto::class, $data))
```
  Implementation:
```php
  public function projectAs(string $dtoClass, ?callable $factory = null): static {
      $this->_dtoClass = $dtoClass;
      $this->_dtoFactory = $factory;
      return $this;
  }
```

markstory commented:
>  ->projectAs(UserDto::class, fn($data) => $valinorMapper->map(UserDto::class, $data))

Is this something that is needed now? I think we should see if we can get the design right and get some adoption before we need to make it pluggable.

dereuromark commented:
Providing the factory possibility wouldn't hurt IMO, since it would not be used in default cases anyway.
Without it using custom DTO libraries or any mapping library would be not easy to achieve I assume.

rochamarcelo commented:
Does this feature works with column type Enum?
```
//In table initialize
 $schema->setColumnType('status', EnumType::from(\App\Enum\Status::class));

//In DTO class
`public \App\Enum\Status $status` 
````

dereuromark commented:
I would think so. Enums are also just classes. The DTO just has to define them. See test cases in cake-dto.

//EDIT @rochamarcelo I added a real life test in sandbox for enums, works.

dereuromark commented:
Are we then able to move forward?

markstory commented on tests/test_app/TestApp/Dto/ArticleWithDatesDto.php:
```diff
@@ -0,0 +1,26 @@
+    public function __construct(
+        public int $id,
+        public string $title,
+        public ?DateTime $created = null,
+        public ?DateTime $modified = null,
```
Do we need special casing for application enums or other custom application type mappings?

dereuromark replied:
As per my sandbox example code this all works as is.

markstory reviewed (commented):
Generally looks good to me. I think this feature will require documentation on what a DTO is, and give developers context on when/why they would use `projectAs()` instead of getting results as entities & arrays.
